File: java/src/leetcode32.java
/**
 * @author laegel
 * @version 1.0
 * @since 2025-10-21
 */
public class leetcode32 {
    public static void main(String[] args) {
        System.out.println(longestValidParentheses(""));
    }

    public static int longestValidParentheses(String s) {
        int maxLength = 0;
        int l = 0;
        int r = 0;

        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                l++;
            } else {
                r++;
            }

            if (l == r) {
                maxLength = Math.max(maxLength, l + r);
            } else if (r > l) {
                l = 0;
                r = 0;
            }
        }

        l = 0;
        r = 0;
        for (int i = s.length() - 1; i >= 0; i--) {
            if (s.charAt(i) == ')') {
                r++;
            } else {
                l++;
            }

            if (l == r) {
                maxLength = Math.max(maxLength, l + r);
            } else if (l > r) {
                l = 0;
                r = 0;
            }
        }

        return maxLength;
    }
}

File: python/leetcode32.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        l, r = 0
        max_len = 0
        for i in range(len(s)):
            if s[i] == "(":
                l = l + 1
            else:
                r = r + 1

            if l == r:
                max_len = max(max_len, l + r)
            elif r > l:
                l = 0
                r = 0

        l = 0
        r = 0
        for i in range(len(s) - 1, -1, -1):
            if s[i] == ")":
                r = r + 1
            else:
                l = l + 1

            if l == r:
                max_len = max(max_len, l + r)
            elif l > r:
                l = 0
                r = 0


        return max_len


